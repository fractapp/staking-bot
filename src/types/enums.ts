export enum Network {
    Polkadot = 0,
    Kusama,
}

export enum Action {
    Open,
    AddAmount,
    CreateWithdrawRq,
    ConfirmWithdraw,
    UpdateNominations
}


export enum Currency {
    DOT = 0,
    KSM,
}


export function getNativeCurrency(network: Network): Currency {
    switch (network) {
        case Network.Polkadot:
            return Currency.DOT
        case Network.Kusama:
            return Currency.KSM
    }
}

export function getNetwork(currency: Currency): Network {
    switch (currency) {
        case Currency.DOT:
            return Network.Polkadot
        case Currency.KSM:
            return Network.Kusama
    }
}

export function getSymbol(currency: Currency) {
    let symbol = '';
    switch (currency) {
        case Currency.DOT:
            symbol = 'DOT';
            break;
        case Currency.KSM:
            symbol = 'KSM';
            break;
        default:
            throw new Error('invalid currency');
    }
    return symbol;
}

export function toCurrency(currency: string): Currency {
    switch (currency) {
        case 'DOT':
            return Currency.DOT;
        case 'KSM':
            return Currency.KSM;
        default:
            throw new Error('invalid currency');
    }
}

export function fromCurrency(currency: Currency): string {
    switch (currency) {
        case Currency.DOT:
            return 'DOT';
        case Currency.KSM:
            return 'KSM';
        default:
            throw new Error('invalid currency');
    }
}
