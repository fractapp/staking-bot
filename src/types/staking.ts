export type ValidatorInfo = {
    ownStake: string,
    totalStake: string,
    commission: number,
    isBlocksNominations: boolean,
    isSlash: boolean
    isOverSubscribed: boolean
    isIdentity: boolean,
}

export type StakingInfo = {
    activeEra: number,
    minStakingAmountPlanks: bigint,
    maxNominations: number,
    maxNominatorRewardedPerValidator: number,
    isAvailable: boolean,
    averageAPY: number,
    decimalsCount: number,
    eraDuration: number,
    decimals: bigint,
    unbondingPeriod: number
}

export type ConstantsInfo = {
    MaxNominations: number,
    MaxNominatorRewardedPerValidator: number,
    ExpectedBlockTime: number,
    EpochDuration: number,
    BondingDuration: number,
}
