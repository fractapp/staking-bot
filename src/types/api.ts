import {Currency} from "./enums";

export type UndeliveredMessagesInfo = {
    messages: Array<Message>,
    users: {
        [id in string]: Profile
    }
};
export type Profile = {
    id: string;
    name: string;
    username: string;
    avatarExt: string;
    lastUpdate: number;
    addresses: {
        [id in Currency]: string
    };
    isChatBot: boolean;
};

export type MessageRq = {
    version: number,
    value: string,
    action: string,
    receiver: string,
    args: {
        [key in string]: string
    },
    rows: Array<Row>
};

export type Message = {
    id: string;
    value: string;
    action: string | null;
    args: {
        [key in string]: string
    },
    rows: Array<Row>;
    timestamp: number;
    sender: string;
    receiver: string;
    hideBtn: boolean;
};

export type Row = {
    buttons: Array<Button>
};

export type Button = {
    value: string
    action: string
    arguments: {
        [key in string]: string
    }
    imageUrl: string | null,
};

export enum DefaultMsgAction {
    Init = "/init",
    OpenUrl = "/openUrl",
    EnterAmount = "/enterAmount",
    Broadcast = "/broadcast",
    WalletButtonOut = '/walletOut',
    WalletButtonIn = '/walletIn'
}
