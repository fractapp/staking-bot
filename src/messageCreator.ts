import {Button, Message, Profile} from "./types/api";
import {Action, Currency, fromCurrency, Network} from "./types/enums";
import {MathUtil} from "./math";
import {ApiPromise} from "@polkadot/api";
import {Cache} from "./cache";
import {Const} from "./const";
import {MsgAction} from "./actions";
import BN from "bn.js";

export type DepositInfo = {
    value: string | null,
    hasWithdrawRequest: boolean,
    hasActiveAmount: boolean,
    isWarnExist: boolean
}

export type NewDepositInfo = {
    value: string,
    hasOpenedDeposit: boolean
}

export type WithdrawRequestsInfo = {
    value: string | null,
    totalWithdrawPlanks: BN,
    buttons: Array<Button>
}

export class MessageCreator {
    private readonly api: ApiPromise
    private readonly cache: Cache
    private readonly network: Network
    private readonly currency: Currency

    constructor(api: ApiPromise, cache: Cache, network: Network, currency: Currency) {
        this.api = api
        this.cache = cache
        this.network = network
        this.currency = currency
    }

    public warning(activeAmount: BN): string | null {
        const cacheInfo = this.cache.getStakingInfo()
        const amount = MathUtil.convertFromPlanckToString(cacheInfo.minStakingAmountPlanks, cacheInfo.decimalsCount)

        if (activeAmount.cmp(cacheInfo.minStakingAmountPlanks) < 0) {
            return `\n\nThe $${fromCurrency(this.currency)} deposit is not active. The deposit must have more than the minimum amount of ${amount} ${fromCurrency(this.currency)}!`
        }

        return null
    }
    public async getNewDeposit(user: Profile): Promise<NewDepositInfo> {
        const info = this.cache.getStakingInfo()
        const min = info.minStakingAmountPlanks.mul(Const.Accuracy).div(info.decimals).toNumber() / Const.Accuracy.toNumber()
        const staking = await this.api.query.staking.ledger(user.addresses[this.currency])
        const stakingInfo = staking.isNone ? null : staking.unwrap()

        const value =
            (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
            `Annual yield: ${info.averageAPY}%\n` +
            `Withdrawal duration: ${info.unbondingPeriod} days \n` +
            `Minimum amount: ${min} ${fromCurrency(this.currency)}`

        return {
            value: value,
            hasOpenedDeposit: stakingInfo != null && stakingInfo.total.toBn().cmp(new BN(0)) != 0
        }
    }

    public async getDeposit(user: Profile): Promise<DepositInfo> {
        const staking = await this.api.query.staking.ledger(user.addresses[this.currency])
        const cacheInfo = this.cache.getStakingInfo()

        const stakingInfo = staking.isNone ? null : staking.unwrap()

        let value = null
        let isWarnExist = false
        if (stakingInfo != null && stakingInfo.total.toBn().cmp(new BN(0)) != 0) {
            const warn = this.warning(stakingInfo.active.toBn())
            if (warn != null) {
                isWarnExist = true
            }

            value =   (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
                `Withdrawal requests: ${stakingInfo!.unlocking.length}\n` +
                `Active amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.active.toBn(), cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}\n` +
                `Withdrawal amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.total.toBn().sub(stakingInfo!.active.toBn()), cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}` +
                (warn ?? "")
        }

        return {
            value: value,
            hasWithdrawRequest: stakingInfo != null && stakingInfo!.unlocking.length > 0,
            hasActiveAmount: stakingInfo != null && stakingInfo!.active.toBn().cmp(new BN(0)) > 0,
            isWarnExist: isWarnExist
        }
    }

    public async getWithdrawRequests(user: Profile): Promise<WithdrawRequestsInfo> {
        const staking = await this.api.query.staking.ledger(user.addresses[this.currency])

        const info = this.cache.getStakingInfo()
        const stakingInfo = staking.isNone ? null : staking.unwrap()

        const buttons = []
        let totalWithdrawPlanks = new BN(0)
        let withdrawRequestText = ""
        if (stakingInfo != null && stakingInfo.unlocking.length > 0) {
            const activeEra = await this.api.query.staking.activeEra()

            withdrawRequestText = this.currency == Currency.DOT ? `ðŸ”´DOT requestsðŸ”´\n\n` : `ðŸ”µKSM requestsðŸ”µ\n\n`

            let isExistNotActiveWR = false
            for (let unlocking of stakingInfo.unlocking) {
                const planksValue = unlocking.value.unwrap().toBn()
                const amount = MathUtil.convertFromPlanckToString(planksValue, info.decimalsCount)
                const era = unlocking.era.unwrap().toNumber()
                const duration = era - activeEra.unwrap().index.toBn().toNumber()
                const time = Math.floor(duration * info.eraDuration) / 24

                if (time < 0) {
                    totalWithdrawPlanks = totalWithdrawPlanks.add(planksValue)
                } else {
                    if (!isExistNotActiveWR) {
                        isExistNotActiveWR = true
                    }
                    withdrawRequestText += `${amount} ${fromCurrency(this.currency)} unlocked after ${time} days\n`
                }
            }

            if (!isExistNotActiveWR) {
                withdrawRequestText = this.warning(stakingInfo.active.toBn()) ?? ""
            } else {
                withdrawRequestText += this.warning(stakingInfo.active.toBn()) ?? ""
            }
        }

        let value: string | null = null
        if (withdrawRequestText) {
            value = withdrawRequestText
        }

        if (totalWithdrawPlanks.cmp(new BN(0)) > 0) {
            const withdrawText = `${MathUtil.convertFromPlanckToString(totalWithdrawPlanks, info.decimalsCount)} ${fromCurrency(this.currency)}`
            value += `\nYou can withdraw ${withdrawText}`
            buttons.push({
                value: `Withdraw ${withdrawText}`,
                action: MsgAction.Confirm,
                arguments: {
                    arguments: JSON.stringify({
                        currency: fromCurrency(this.currency),
                        action: String(Action.ConfirmWithdraw)
                    })
                },
                imageUrl: null
            })
        }

        if (stakingInfo != null && stakingInfo.active.toBn().cmp(new BN(0)) > 0) {
            buttons.push({
                value: `Create withdraw request for ${fromCurrency(this.currency)}`,
                action: MsgAction.Enter,
                arguments: {
                    currency: fromCurrency(this.currency),
                    action: String(Action.CreateWithdrawRq)
                },
                imageUrl: null
            })
        }

        return {
            value: value,
            totalWithdrawPlanks: totalWithdrawPlanks,
            buttons: buttons,
        }
    }
}
