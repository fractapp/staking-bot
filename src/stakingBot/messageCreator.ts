import {Button, Profile} from "../types/api";
import {Action, Currency, fromCurrency, MsgAction, Network} from "../types/enums";
import {MathUtil} from "../utils/math";
import {Const} from "../utils/const";
import {CacheClient} from "../utils/cacheClient";

export type DepositInfo = {
    value: string | null,
    hasWithdrawRequest: boolean,
    hasActiveAmount: boolean,
    isWarnExist: boolean
}

export type NewDepositInfo = {
    value: string,
    hasOpenedDeposit: boolean
}

export type WithdrawRequestsInfo = {
    value: string | null,
    totalWithdrawPlanks: bigint,
    buttons: Array<Button>
}

export class MessageCreator {
    private readonly cache: CacheClient
    private readonly network: Network
    private readonly currency: Currency

    constructor(cache: CacheClient, network: Network, currency: Currency) {
        this.cache = cache
        this.network = network
        this.currency = currency
    }

    public async warning(activeAmount: bigint): Promise<string | null> {
        const cacheInfo = await this.cache.getStakingInfo()

        if (activeAmount < cacheInfo.minStakingAmountPlanks) {
            const minAmount = MathUtil.convertFromPlanckToString(cacheInfo.minStakingAmountPlanks, cacheInfo.decimalsCount)
            return `\n\nThe ${fromCurrency(this.currency)} deposit is not active. The deposit must have more than the minimum amount of ${minAmount} ${fromCurrency(this.currency)}!`
        }

        return null
    }

    public async getNewDeposit(user: Profile): Promise<NewDepositInfo> {
        const info = await this.cache.getStakingInfo()
        const usersStaking = await this.cache.getUsersStaking()
        const min = Number(info.minStakingAmountPlanks * Const.Accuracy / info.decimals) / Number(Const.Accuracy)
        const staking = usersStaking[user.addresses[this.currency]]
        const stakingInfo = staking == undefined ? null : staking

        const value =
            (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
            `Annual yield: ${info.averageAPY}%\n` +
            `Withdrawal duration: ${info.unbondingPeriod} days \n` +
            `Minimum amount: ${min} ${fromCurrency(this.currency)}`

        return {
            value: value,
            hasOpenedDeposit: stakingInfo != null && stakingInfo.total != 0n
        }
    }

    public async getDeposit(user: Profile): Promise<DepositInfo> {
        const cacheInfo = await this.cache.getStakingInfo()
        const usersStaking = await this.cache.getUsersStaking()

        const staking = usersStaking[user.addresses[this.currency]]
        const stakingInfo = staking == undefined ? null : staking

        let value = null
        let isWarnExist = false
        if (stakingInfo != null && stakingInfo.total != 0n) {
            const warn = await this.warning(stakingInfo.active)
            if (warn != null) {
                isWarnExist = true
            }

            value = (this.currency == Currency.KSM ? 'ðŸ”µKSM DepositðŸ”µ\n' : 'ðŸ”´DOT DepositðŸ”´\n') +
                `Total withdrawal request: ${stakingInfo!.unlocking.length}\n` +
                `Active deposit amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.active, cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}\n` +
                `Withdrawal amount: ${MathUtil.convertFromPlanckToString(stakingInfo!.total - stakingInfo!.active, cacheInfo.decimalsCount)} ${fromCurrency(this.currency)}` +
                (warn ?? "")
        }

        return {
            value: value,
            hasWithdrawRequest: stakingInfo != null && stakingInfo!.unlocking.length > 0,
            hasActiveAmount: stakingInfo != null && stakingInfo!.active > 0,
            isWarnExist: isWarnExist
        }
    }

    public async getWithdrawRequests(user: Profile): Promise<WithdrawRequestsInfo> {
        const info = await this.cache.getStakingInfo()
        const usersStaking = await this.cache.getUsersStaking()

        const staking = usersStaking[user.addresses[this.currency]]
        const stakingInfo = staking == undefined ? null : staking

        const buttons = []
        let totalWithdrawPlanks = 0n
        let withdrawRequestText = ""
        if (stakingInfo != null && stakingInfo.unlocking.length > 0) {
            const activeEra = info.activeEra

            withdrawRequestText = this.currency == Currency.DOT ? `ðŸ”´DOT requestsðŸ”´\n\n` : `ðŸ”µKSM requestsðŸ”µ\n\n`

            let isExistNotActiveWR = false
            for (let unlocking of stakingInfo.unlocking) {
                const planksValue = unlocking.value
                const amount = MathUtil.convertFromPlanckToString(planksValue, info.decimalsCount)
                const duration = unlocking.era - activeEra
                const time = Math.floor(duration * info.eraDuration) / 24

                if (time < 0) {
                    totalWithdrawPlanks += planksValue
                } else {
                    if (!isExistNotActiveWR) {
                        isExistNotActiveWR = true
                    }
                    withdrawRequestText += `${amount} ${fromCurrency(this.currency)} unlocked after ${time} days\n`
                }
            }

            if (!isExistNotActiveWR) {
                withdrawRequestText = await this.warning(stakingInfo.active) ?? ""
            } else {
                withdrawRequestText += await this.warning(stakingInfo.active) ?? ""
            }
        }

        let value: string | null = null
        if (withdrawRequestText) {
            value = withdrawRequestText
        }

        if (totalWithdrawPlanks > 0n) {
            const withdrawText = `${MathUtil.convertFromPlanckToString(totalWithdrawPlanks, info.decimalsCount)} ${fromCurrency(this.currency)}`
            value += `\n\nYou can withdraw ${withdrawText}`
            buttons.push({
                value: `Withdraw ${withdrawText}`,
                action: MsgAction.Confirm,
                arguments: {
                    arguments: JSON.stringify({
                        currency: fromCurrency(this.currency),
                        action: String(Action.ConfirmWithdraw)
                    })
                },
                imageUrl: null
            })
        }

        if (stakingInfo != null && stakingInfo.active > 0n) {
            buttons.push({
                value: `Create withdraw request for ${fromCurrency(this.currency)}`,
                action: MsgAction.Enter,
                arguments: {
                    currency: fromCurrency(this.currency),
                    action: String(Action.CreateWithdrawRq)
                },
                imageUrl: null
            })
        }

        return {
            value: value,
            totalWithdrawPlanks: totalWithdrawPlanks,
            buttons: buttons,
        }
    }
}
