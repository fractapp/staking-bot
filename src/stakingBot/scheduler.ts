import {DB} from "../db/db";
import {Action, fromCurrency, getNativeCurrency, Network} from "../types/enums";
import {ActionHandler} from "./actions";
import {Message, Profile} from "../types/api";
import {Const} from "../utils/const";
import {CacheClient} from "../utils/cacheClient";
import {UserStakingInfo} from "../cache/cache";

export class Scheduler  {
    private cacheByNetwork = new Map<Network, CacheClient>()

    private db: DB
    private actionHandler: ActionHandler;

    constructor(
        cachePolkadot: CacheClient,
        cacheKusama: CacheClient,
        db: DB,
        actionHandler: ActionHandler
    ) {
        this.db = db

        this.cacheByNetwork.set(Network.Polkadot, cachePolkadot)
        this.cacheByNetwork.set(Network.Kusama, cacheKusama)

        this.actionHandler = actionHandler
    }

    public async call() {
        const usersStakingInfoByNetwork = new Map<Network, Record<string, UserStakingInfo | undefined>>()
        usersStakingInfoByNetwork.set(Network.Polkadot, await this.cacheByNetwork.get(Network.Polkadot)!.getUsersStaking())
        usersStakingInfoByNetwork.set(Network.Kusama, await this.cacheByNetwork.get(Network.Kusama)!.getUsersStaking())

        const sentWithdrawRqByUsers = new Map<string, boolean>()

        const users = await this.db.getUsers()
        for (const user of users) {
            console.log("user: " + JSON.stringify(user))

            for (const network of [Network.Polkadot, Network.Kusama]) {
                const topValidators = await this.cacheByNetwork.get(network)!.getTopValidators()

                const currency = getNativeCurrency(network)
                const usersStakingInfo = usersStakingInfoByNetwork.get(network)!

                const address = user.addresses[network]
                const stakingInfo = usersStakingInfo[user.addresses[network]]
                if (stakingInfo == undefined) {
                    continue
                }

                try {
                    if (stakingInfo.total != stakingInfo.active && stakingInfo.hasWithdrawal && !sentWithdrawRqByUsers.has(user.userId)) {
                        sentWithdrawRqByUsers.set(user.userId, true)

                        console.log(`send withdrawal request (${address}): ${user.userId}`)
                        await this.actionHandler.withdrawRequests(<Message>{
                            sender: user.userId,
                        }, <Profile>{
                            addresses: user.addresses
                        })
                    }

                    const topValidatorsMap = new Map<string, boolean>()
                    for (let validator of topValidators) {
                        topValidatorsMap.set(validator, true)
                    }

                    if (stakingInfo.active == 0n) {
                        continue
                    }

                    let count = 0
                    for (let nomination of stakingInfo.targets) {
                        if (topValidatorsMap.has(nomination)) {
                            count++
                        }
                    }

                    if (
                        (count < topValidators.length && topValidators.length/count >  Const.ThresholdUserNominations) ||
                        (count / topValidators.length <= Const.ThresholdUserNominations)
                    ) {
                        console.log(`send update nominations (${address}): ${user.userId}`)
                        const args: Record<string, string> = {
                            arguments: JSON.stringify({
                                action: String(Action.UpdateNominations),
                                currency: fromCurrency(currency)
                            })
                        }

                        await this.actionHandler.confirm(<Message>{
                            sender: user.userId,
                            args: args
                        }, <Profile>{
                            addresses: user.addresses
                        })
                    }
                } catch (e) {
                    console.log("Error: " + (e as Error).toString())
                }
            }
        }
    }
}
