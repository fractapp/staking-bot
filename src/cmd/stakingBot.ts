import {DefaultMsgAction, Message, Profile} from "../types/api";
import {MsgAction, Network, toCurrency} from "../types/enums";
import {ApiPromise, WsProvider} from "@polkadot/api";
import dotenv from "dotenv";
import {ActionHandler} from "../stakingBot/actions";
import {FractappClient} from "../fractapp/client";
import {DB} from "../db/db";
import {Scheduler} from "../stakingBot/scheduler";
import {CacheClient} from "../utils/cacheClient";
import {Const} from "../utils/const";

dotenv.config()

const clientUrl = process.env["FRACTAPP_CLIENT_URL"] as string
const client = new FractappClient(clientUrl)
let actionHandler: ActionHandler

let cache: Map<Network, CacheClient> = new Map<Network, CacheClient>()

async function start() {
    const connectionString = process.env["MONGODB_CONNECTION"] as string
    const cacheServerPort = process.env["CACHE_PORT"] as string

    const seed = process.env["SEED"] as string
    const apiPolkadot = await getApiInstance(process.env["POLKADOT_RPC_URL"] as string)
    const apiKusama = await getApiInstance(process.env["KUSAMA_RPC_URL"] as string)

    cache.set(Network.Polkadot, new CacheClient(cacheServerPort, Network.Polkadot))
    cache.set(Network.Kusama, new CacheClient(cacheServerPort, Network.Kusama))

    const db = new DB(connectionString)

    console.log("connect to db...")
    await db.connect()
    console.log("end of connect to db")

    actionHandler = new ActionHandler(client, apiPolkadot, apiKusama, cache.get(Network.Polkadot)!, cache.get(Network.Kusama)!, db)
    const scheduler = new Scheduler(cache.get(Network.Polkadot)!, cache.get(Network.Kusama)!, db, actionHandler)

    console.log("init blockchain info...")

    let isInit = false
    while (!isInit) {
        const polkadotCache = cache.get(Network.Polkadot)!
        const kusamaCache = cache.get(Network.Kusama)!

        console.log("init cache")

        try {
            await polkadotCache.updateCache()
            await kusamaCache.updateCache()

            await actionHandler.init()

            isInit = true
        } catch (e) {
            console.log("wait init cache")
        }

        await new Promise(resolve => setTimeout(resolve, 5 * Const.Sec));
    }

    console.log("end of init blockchain info")

    console.log("fractapp authorization...")
    await client.auth(seed)
    console.log("fractapp authorization: Success")

    setInterval(async () => {
        console.log("start scheduler")
        await scheduler.call()
        console.log("end of scheduler")
    }, Const.SchedulerTimeout)

    console.log("start app...")
    const usersPromise: Map<string, Promise<void>> = new Map<string, Promise<void>>()

    while (true) {
        try {
            const newMsgs = await client.getUnreadMessages()

            for (const msg of newMsgs.messages) {
                console.log("msg id: " + msg.id)
                console.log("msg: " + JSON.stringify(msg))

                if (!usersPromise.has(msg.sender)) {
                    try {
                        const p = action(msg, newMsgs.users[msg.sender]).finally(() => usersPromise.delete(msg.sender))
                        usersPromise.set(msg.sender, p)
                    } catch (e) {
                        usersPromise.delete(msg.sender)
                        console.log("Error: " + e)
                    }
                }
            }
        } catch (e) {
            console.log("Error: " + e)
        }
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
}

async function getApiInstance(wssUrl: string): Promise<ApiPromise> {
    const wsProvider = new WsProvider(wssUrl);
    return await ApiPromise.create({provider: wsProvider})
}

async function action(msg: Message, user: Profile) {
    //TODO: error count for msg
    try {
        switch (msg.action) {
            case DefaultMsgAction.Init:
                await actionHandler.initMsg(msg, user)
                break
            case MsgAction.ChooseNewDeposit:
                await actionHandler.chooseNewDeposit(msg, user)
                break
            case MsgAction.Enter:
                await actionHandler.enter(toCurrency(msg.args.currency), Number(msg.args.action), msg.sender, user)
                break
            case MsgAction.ChooseWithdraw:
                await actionHandler.chooseWithdraw(msg)
                break
            case MsgAction.Confirm:
                await actionHandler.confirm(msg, user)
                break
            case MsgAction.SuccessTx:
                await actionHandler.successTx(msg)
                break
            case MsgAction.ErrorTx:
                await actionHandler.errorTx(msg)
                break
            case DefaultMsgAction.WalletButtonIn:
            case MsgAction.MyDeposits:
                await actionHandler.myDeposits(msg, user)
                break
            case DefaultMsgAction.WalletButtonOut:
            case MsgAction.WithdrawRequests:
                await actionHandler.withdrawRequests(msg, user)
                break
            default:
                await actionHandler.initMsg(msg, user)
        }
    } catch (e) {
        await actionHandler.initMsg(msg, user)
    }
    await client.setRead([msg.id])
    console.log("Success: (msgId)" + msg.id)
}

start()
