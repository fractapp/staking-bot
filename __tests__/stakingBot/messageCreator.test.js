import {Currency, Network} from "../../src/types/enums";
import {MessageCreator} from "../../src/stakingBot/messageCreator";

const cache = ({
    getStakingInfo: jest.fn(),
    getUsersStaking: jest.fn()
})
const messageCreator = new MessageCreator(cache, Network.Polkadot, Currency.DOT)

it('test warning', async () => {
    cache.getStakingInfo.mockReturnValue({
        minStakingAmountPlanks: 123n,
        decimalsCount: 3
    })
    expect(await messageCreator.warning(1234n)).toEqual(null)
    expect(await messageCreator.warning(13n)).toMatchSnapshot()
});

it('test getNewDeposit', async () => {
    cache.getStakingInfo.mockReturnValueOnce({
        minStakingAmountPlanks: 1234n,
        decimals: 1000n
    })
    cache.getUsersStaking.mockReturnValueOnce({})

    const r = await messageCreator.getNewDeposit({
        addresses: {
            0: "addressDOT"
        }
    })
    expect(r.value).toMatchSnapshot()
    expect(r.hasOpenedDeposit).toEqual(false)
});

it('test getDeposit with non-empty deposit', async () => {
    cache.getStakingInfo.mockReturnValue({
        minStakingAmountPlanks: 123n,
        decimalsCount: 3,
        decimals: 1000n,
    })
    cache.getUsersStaking.mockReturnValueOnce({
        "addressDOT": {
            unlocking: [ "1", "2" ],
            active: 900n,
            total: 1000n,
        }
    })

    const r = await messageCreator.getDeposit({
        addresses: {
            0: "addressDOT"
        }
    })
    expect(r.value).toMatchSnapshot()
    expect(r.hasWithdrawRequest).toEqual(true)
    expect(r.hasActiveAmount).toEqual(true)
    expect(r.isWarnExist).toEqual(false)
});

it('test getDeposit with empty deposit', async () => {
    cache.getStakingInfo.mockReturnValue({
        minStakingAmountPlanks: 123n,
        decimalsCount: 3,
        decimals: 1000n,
    })
    cache.getUsersStaking.mockReturnValueOnce({
    })

    const r = await messageCreator.getDeposit({
        addresses: {
            0: "addressDOT"
        }
    })
    expect(r.value).toEqual(null)
    expect(r.hasWithdrawRequest).toEqual(false)
    expect(r.hasActiveAmount).toEqual(false)
    expect(r.isWarnExist).toEqual(false)
});

it('test getWithdrawRequests', async () => {
    cache.getStakingInfo.mockReturnValue({
        activeEra: 100,
        minStakingAmountPlanks: 123n,
        decimalsCount: 3,
        decimals: 1000n,
        eraDuration: 5
    })
    cache.getUsersStaking.mockReturnValueOnce({
        "addressDOT": {
            unlocking: [
                {
                    value: 1250n,
                    era: 101
                },
                {
                    value: 2250n,
                    era: 99
                },
                {
                    value: 3250n,
                    era: 99
                },
                {
                    value: 250n,
                    era: 101
                }
            ],
            active: 900n,
            total: 1000n,
        }
    })

    const r = await messageCreator.getWithdrawRequests({
        addresses: {
            0: "addressDOT"
        }
    })
    expect(r.value).toMatchSnapshot()
    expect(r.buttons).toEqual([
        {
            action: "confirm",
            arguments: {
                arguments: "{\"currency\":\"DOT\",\"action\":\"3\"}",
            },
            imageUrl: null,
            value: "Withdraw 5.5 DOT",
        },
        {
            action: "enter",
            arguments: {
                action: "2",
                currency: "DOT"
            },
            imageUrl: null,
            value:  "Create withdraw request for DOT",
        }
    ])
    expect(r.totalWithdrawPlanks).toEqual(5500n)
});
