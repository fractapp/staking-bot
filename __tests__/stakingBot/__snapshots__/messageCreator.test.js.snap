// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test getDeposit with non-empty deposit 1`] = `
"ðŸ”´DOT DepositðŸ”´
Total withdrawal request: 2
Active deposit amount: 0.9 DOT
Withdrawal amount: 0.1 DOT"
`;

exports[`test getNewDeposit 1`] = `
"ðŸ”´DOT DepositðŸ”´
Annual yield: undefined%
Withdrawal duration: undefined days 
Minimum amount: 1.234 DOT"
`;

exports[`test getWithdrawRequests 1`] = `
"ðŸ”´DOT requestsðŸ”´

1.25 DOT unlocked after 0.20833333333333334 days
0.25 DOT unlocked after 0.20833333333333334 days


You can withdraw 5.5 DOT"
`;

exports[`test warning 1`] = `
"

The DOT deposit is not active. The deposit must have more than the minimum amount of 0.123 DOT!"
`;
